File: haskell/packages/translator/tests/TranslatorTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Data.Time.Clock (getCurrentTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Test.Tasty (TestTree, defaultMain, testGroup)
import Test.Tasty.HUnit (assertBool, testCase)

import Translator (translate)

-- |
-- = @formatErrorMessage@
--
-- The function @formatErrorMessage@ is used to format the error message.
--
-- __Arguments:__
--
-- * @expected: String@ - The expected output.
-- * @actual: String@ - The actual output.
--
-- __Return:__
--
-- * @IO String@ - The formatted error message.
formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

-- |
-- = @translateTests@
--
-- The @translateTests@ function contains the tests for the @translate@
-- function.
translateTests :: TestTree
translateTests =
    let name = "Alter Sabeh"
        languages =
            [ ("English", "Hi Alter Sabeh, greetings from Haskell !")
            , ("French", "Bonjour Alter Sabeh, salutations de Haskell !")
            , ("Spanish", "Hola Alter Sabeh, saludos desde Haskell !")
            , ("German", "Hallo Alter Sabeh, grüße aus Haskell !")
            , ("Italian", "Ciao Alter Sabeh, saluti da Haskell !")
            , ("Dutch", "Hallo Alter Sabeh, groeten uit Haskell !")
            , ("Unknown", "Hi Alter Sabeh, greetings from Haskell !")
            ]
     in testGroup "Language" $
            map
                ( \(lang, expectedOutput) ->
                    testCase lang $ do
                        let actualOutput = translate lang name
                        errorMessage <- formatErrorMessage expectedOutput actualOutput
                        assertBool errorMessage (expectedOutput == actualOutput)
                )
                languages

-- |
-- = @tests@
--
-- This function contains the tests to be run.
tests :: TestTree
tests = testGroup "Translator Tests" [translateTests]

-- |
-- = @main@
--
-- The @main@ function is the entrypoint of the program.
main :: IO ()
main = defaultMain tests

File: kotlin/subprojects/translator/bin/test/com/app/translator/TranslatorTests.kt
package com.prog.translator

import org.apache.commons.lang3.time.DateFormatUtils
import kotlin.test.Test
import kotlin.test.assertEquals

class TranslatorTests {
    /**
     * # `formatErrorMessage`
     *
     * This function formats the error message for the test.
     *
     * __Arguments:__
     *
     * * `expected: String` - The expected output.
     * * `actual: String` - The actual output.
     *
     * __Return:__
     *
     * * `String` - The formatted error message.
     */
    private fun formatErrorMessage(
        expected: String,
        actual: String,
    ): String {
        val now = System.currentTimeMillis()
        val formattedDate = DateFormatUtils.format(now, "yyyy-MM-dd'T'HH:mm:ss")
        return """
            |${"\u001B[31m"}
            |Test run at: $formattedDate
            |Expected : $expected
            |  Actual : $actual
            |
            |${"\u001B[00m"}
            """.trimMargin()
    }

    /**
     * # `translate ENGLISH test`
     *
     * This function tests the `translate` function with English language.
     */
    @Test
    fun `translate ENGLISH test`() {
        val lang = "English"
        val expectedOutput = "Hi Alter Sabeh, greetings from Kotlin !"
        val name = "Alter Sabeh"
        val actualOutput = translate(lang, name)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `translate FRENCH test`
     *
     * This function tests the `translate` function with French language.
     */
    @Test
    fun `translate FRENCH test`() {
        val lang = "French"
        val expectedOutput = "Bonjour Alter Sabeh, salutations de Kotlin !"
        val name = "Alter Sabeh"
        val actualOutput = translate(lang, name)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `translate SPANISH test`
     *
     * This function tests the `translate` function with Spanish language.
     */
    @Test
    fun `translate SPANISH test`() {
        val lang = "Spanish"
        val expectedOutput = "Hola Alter Sabeh, saludos desde Kotlin !"
        val name = "Alter Sabeh"
        val actualOutput = translate(lang, name)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `translate GERMAN test`
     *
     * This function tests the `translate` function with German language.
     */
    @Test
    fun `translate GERMAN test`() {
        val lang = "German"
        val expectedOutput = "Hallo Alter Sabeh, grüße aus Kotlin !"
        val name = "Alter Sabeh"
        val actualOutput = translate(lang, name)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `translate ITALIAN test`
     *
     * This function tests the `translate` function with Italian language.
     */
    @Test
    fun `translate ITALIAN test`() {
        val lang = "Italian"
        val expectedOutput = "Ciao Alter Sabeh, saluti da Kotlin !"
        val name = "Alter Sabeh"
        val actualOutput = translate(lang, name)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `translate DUTCH test`
     *
     * This function tests the `translate` function with Dutch language.
     */
    @Test
    fun `translate DUTCH test`() {
        val lang = "Dutch"
        val expectedOutput = "Hallo Alter Sabeh, groeten uit Kotlin !"
        val name = "Alter Sabeh"
        val actualOutput = translate(lang, name)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }

    /**
     * # `translate UNKNOWN test`
     *
     * This function tests the `translate` function with unknown language.
     */
    @Test
    fun `translate UNKNOWN test`() {
        val lang = "Unknown"
        val expectedOutput = "Hi Alter Sabeh, greetings from Kotlin !"
        val name = "Alter Sabeh"
        val actualOutput = translate(lang, name)
        val errorMessage = formatErrorMessage(expectedOutput, actualOutput)
        assertEquals(expectedOutput, actualOutput, errorMessage)
    }
}
